src/Microsoft.AspNetCore.Sockets.Http/Transports/WebSocketsTransport.cs

```diff
@@ -126,69 +117,93 @@ public WebSocketsTransport(WebSocketOptions options, IChannelConnection<Message>
             }
         }
 
-        private Task HandleFrame(WebSocketFrame frame)
+        private async Task<WebSocketReceiveResult> StartReceiving(WebSocket socket)
         {
-            // Is this a frame we care about?
-            if (!frame.Opcode.IsMessage())
+            // REVIEW: This code was copied from the client, it's highly unoptimized at the moment (especially 
+            // for server logic)
+            var incomingMessage = new List<ArraySegment<byte>>();
+            while (true)
             {
-                return Task.CompletedTask;
-            }
+                const int bufferSize = 4096;
+                var totalBytes = 0;
+                WebSocketReceiveResult receiveResult;
+                do
+                {
+                    var buffer = new ArraySegment<byte>(new byte[bufferSize]);
 
-            LogFrame("Receiving", frame);
+                    // Exceptions are handled above where the send and receive tasks are being run.
+                    receiveResult = await socket.ReceiveAsync(buffer, CancellationToken.None);
```

Comment:
We should enforce that after the first frame of a message, all remaining frames **must** be of type `Continuation` or `Close`. Receiving `Text (Fin=false)` followed by `Text (Fin=true)` is a protocol error (the second message must be `Continuation`) and we should close the socket when that occurs, with status code `1002` (Protocol Error)